libs/langchain-mongodb/langchain_mongodb/vectorstores.py

```diff
@@ -396,24 +431,21 @@ def bulk_embed_and_insert_texts(
             return []
         # Compute embedding vectors
         embeddings = self._embedding.embed_documents(list(texts))
-        if ids:
-            to_insert = [
-                {
-                    "_id": str_to_oid(i),
-                    self._text_key: t,
-                    self._embedding_key: embedding,
-                    **m,
-                }
-                for i, t, m, embedding in zip(ids, texts, metadatas, embeddings)
-            ]
-        else:
-            to_insert = [
-                {self._text_key: t, self._embedding_key: embedding, **m}
-                for t, m, embedding in zip(texts, metadatas, embeddings)
-            ]
+        if not ids:
+            ids = [ObjectId() for _ in range(len(texts))]
+        docs = [
+            {
+                "_id": str_to_oid(i),
+                self._text_key: t,
+                self._embedding_key: embedding,
+                **m,
+            }
+            for i, t, m, embedding in zip(ids, texts, metadatas, embeddings)
+        ]
+        operations = [ReplaceOne({"_id": doc["_id"]}, doc, upsert=True) for doc in docs]
         # insert the documents in MongoDB Atlas
-        insert_result = self._collection.insert_many(to_insert)  # type: ignore
```

Comment:
Does this change cause a performance hit? bulk_write(ReplaceOne) vs insert_many?

What was going here that the ids behaviour had to change?

Reply:
There wasn't a test for that case? Ouch.